algosdk/atomic_transaction_composer.py

```diff
@@ -165,67 +188,156 @@ def add_method_call(
             raise error.AtomicTransactionComposerError(
                 "AtomicTransactionComposer cannot exceed MAX_GROUP_SIZE transactions"
             )
+        if app_id == 0:
+            if not approval_program or not clear_program:
+                raise error.AtomicTransactionComposerError(
+                    "One of the following required parameters for application creation is missing: approvalProgram, clearProgram"
+                )
+        elif on_complete == transaction.OnComplete.UpdateApplicationOC:
+            if not approval_program or not clear_program:
+                raise error.AtomicTransactionComposerError(
+                    "One of the following required parameters for OnApplicationComplete.UpdateApplicationOC is missing: approvalProgram, clearProgram"
+                )
+            if local_schema or global_schema or extra_pages:
+                raise error.AtomicTransactionComposerError(
+                    "One of the following application creation parameters were set on a non-creation call: numGlobalInts, numGlobalByteSlices, numLocalInts, numLocalByteSlices, extraPages"
+                )
+        elif (
+            approval_program
+            or clear_program
+            or local_schema
+            or global_schema
+            or extra_pages
+        ):
+            raise error.AtomicTransactionComposerError(
+                "One of the following application creation parameters were set on a non-creation call: approvalProgram, clearProgram, numGlobalInts, numGlobalByteSlices, numLocalInts, numLocalByteSlices, extraPages"
+            )
         if not method_args:
             method_args = []
         if len(method.args) != len(method_args):
             raise error.AtomicTransactionComposerError(
                 "number of method arguments do not match the method signature"
             )
-        if not isinstance(method, abi.method.Method):
+        if not isinstance(method, abi.Method):
             raise error.AtomicTransactionComposerError(
                 "invalid Method object was passed into AtomicTransactionComposer"
             )
-        if app_id == 0:
-            raise error.AtomicTransactionComposerError(
-                "application create call not supported"
-            )
+
+        # Initialize foreign object maps
+        if not accounts:
+            accounts = []
+        if not foreign_apps:
+            foreign_apps = []
+        if not foreign_assets:
+            foreign_assets = []
```

Comment:
It would be best to make a copy of these input arrays always, since this function may append to them.